api/server/router/system/system_routes.go

```diff
@@ -90,36 +91,77 @@ func (s *systemRouter) getVersion(ctx context.Context, w http.ResponseWriter, r
 }
 
 func (s *systemRouter) getDiskUsage(ctx context.Context, w http.ResponseWriter, r *http.Request, vars map[string]string) error {
-	eg, ctx := errgroup.WithContext(ctx)
-
-	var du *types.DiskUsage
-	eg.Go(func() error {
-		var err error
-		du, err = s.backend.SystemDiskUsage(ctx)
+	if err := httputils.ParseForm(r); err != nil {
 		return err
-	})
+	}
 
-	var buildCache []*types.BuildCache
-	eg.Go(func() error {
+	var getContainers, getImages, getVolumes, getBuildCache bool
+	typesStrs, ok := r.Form["types"]
+	if !ok {
+		getContainers, getImages, getVolumes, getBuildCache = true, true, true, true
+	} else {
+		for _, s := range typesStrs {
+			for _, typ := range strings.Split(s, ",") {
+				switch typ {
+				case "container", "containers":
+					getContainers = true
+				case "image", "images":
+					getImages = true
+				case "volume", "volumes":
+					getVolumes = true
+				case "build", "build-cache", "build-caches":
+					getBuildCache = true
+				default:
+					return pkgerrors.Errorf("unknown object type: %s", typ)
+				}
+			}
+		}
+	}
+
+	var du *types.DiskUsage
+	diskUsage := func() error {
 		var err error
-		buildCache, err = s.builder.DiskUsage(ctx)
+		du, err = s.backend.SystemDiskUsage(ctx, DiskUsageOptions{
+			Containers: getContainers,
+			Images:     getImages,
+			Volumes:    getVolumes,
+		})
 		if err != nil {
-			return pkgerrors.Wrap(err, "error getting build cache usage")
+			return pkgerrors.Wrap(err, "error getting system disk usage")
```

Comment:
Note: a wrapper was missing here, added

Reply:
Perhaps (the "system" in this function's name / route is because it's a `docker system df` command, but perhaps "system disk" may cause some confusion).

```suggestion
			return pkgerrors.Wrap(err, "error getting disk usage")
```

That said, we should look if it might have been on purpose, and if the extra string adds sufficient value; there's some inconsistencies in `SystemDiskUsage` itself as well; some errors are not wrapped (but can return a "typed" error), other's are discarding underlying errors by creating a new error (without `Wrap` or `%w`); https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/disk_usage.go#L15
https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/disk_usage.go#L25
https://github.com/moby/moby/blob/7b9275c0da707b030e62c96b679a976f31f929d3/daemon/disk_usage.go#L31

So this would look (for the above) like:

```
error getting system disk usage: a disk usage operation is already running
error getting system disk usage: failed to retrieve container list: <underlyihg error message>
error getting system disk usage: failed to retrieve image list: <underlyihg error message>
```


The "failed to retrieve container list" and "failed to retrieve image list" cases may actually be wrong, as (looking at the "images" case, the underlying errors may be of a specific `errdef` type; https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/daemon/images/images.go#L53-L61

https://github.com/moby/moby/blob/1ba54a5fd0ba293db3bea46cd67604b593f2048b/api/types/filters/parse.go#L250-L256

Reply:
I think it's better to dive into this in another PR, removing

Reply:
Yes, agreed; should be looked into, but let's take the "safe route" for now